cpp/src/arrow/stl_test.cc

```diff
@@ -49,9 +61,55 @@ struct CustomType {
 #undef ARROW_CUSTOM_TYPE_TIED
 };
 
+// Mock optional object returning null, "yes", "no", null, "yes", "no", ...
```

Comment:
I know this is a test class, but it's a headscrew to me that a single instance of it will sometimes evaluate to true and sometimes to false:

```c++
CustomOptionalTypeMock o;
std::cout << bool(c) << std::endl; // true
std::cout << bool(c) << std::endl; // true
std::cout << bool(c) << std::endl; // false
std::cout << bool(c) << std::endl; // true
// ...
```

It seems to me that in order to emulate an optional, one of these should be consistently null or not null

Reply:
This was actually tailored to be used with `NullableTypesDoNotBreakUserSpecialization` test below. It doesn't entirely mock optional. It checks just that existing user specializations aren't broken. It will be casted in [below](https://github.com/apache/arrow/pull/5171/files#diff-0b35fdf968ac99e27c2c253dfa148f59R105) only. So, it presumes that casting will be done once for each element in a `std::vector<CustomOptionalTypeMock>`.

I can add some comments here clarifying this assumption. Alternatively I can implement a mock iterable of mock optionals to make it consistent. I'd prefer the former, but the latter is okay for me as well.

Reply:
If you want to demonstrate that these objects are cast only once, then please implement them with an initially false boolean member `was_cast_` or something equally unambiguous. Comments would also be welcome

Reply:
Done.